### PR TITLE
UIREC-321: add bind pieces full screen page

### PR DESCRIPTION
**Pull request by alisher-epam**
## Purpose

- [UIREC-321](https://folio-org.atlassian.net/browse/UIREC-321): add bind pieces full screen page
- [UIREC-353](https://folio-org.atlassian.net/browse/UIREC-353): Implement binding button behavior.
- [UIREC-354](https://folio-org.atlassian.net/browse/UIREC-354): Transfer requests for items to be bound
## Approach

https://github.com/folio-org/ui-receiving/assets/116072773/e3c18187-61d1-4dfd-893c-ea7b89acf7cf

## Pre-Merge Checklist
Before merging this PR, please go through the following list and take appropriate actions.
- [ ] I've added appropriate record to the CHANGELOG.md
- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] If any API-related changes - okapi interfaces and permissions are reviewed/changed correspondingly
  - [ ] There are no breaking changes in this PR.

If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

**Review by OleksandrHladchenko1 — APPROVED** [review]

File: src/TitleBindPieces/TitleBindPiecesContainer.test.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+import { useTitle } from '../common/hooks';
+import {
+  usePOLine,
+  useUnboundPieces,
+} from './hooks';
+import TitleBindPieces from './TitleBindPieces';
+import TitleBindPiecesContainer from './TitleBindPiecesContainer';
+
+jest.mock('@folio/stripes/core', () => ({
```
Comment (usavkov-epam) [inline src/TitleBindPieces/TitleBindPiecesContainer.test.js:16]: [Possibly you don't need it](https://github.com/folio-org/ui-receiving/pull/534/files#diff-9afaa1ba5750c66453763dace92ff7ffed804e0f3baa1201594a4f87ddf9b399R63)

File: src/TitleBindPieces/TitleBindPiecesCreateItemForm/TitleBindPiecesCreateItemForm.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+function buildOptions(items) {
+  return items.map(({ id, name }) => ({
+    label: name,
+    value: id,
+  }));
+}
+
+export const TitleBindPiecesCreateItemForm = ({ onChange }) => {
+  const { materialTypes } = useMaterialTypes();
+  const { loanTypes } = useLoanTypes();
```
Comment (usavkov-epam) [inline src/TitleBindPieces/TitleBindPiecesCreateItemForm/TitleBindPiecesCreateItemForm.js:26]: Will a user somehow understand frpm the view that data is loading?

File: src/TitleBindPieces/TitleBindPiecesCreateItemForm/TitleBindPiecesCreateItemForm.js [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+import {
+  Col,
+  Row,
+  Select,
+  TextField,
+} from '@folio/stripes/components';
+import { LocationLookup, LocationSelection } from '@folio/stripes/smart-components';
+
+import { TITLE_BIND_PIECES_FILED_NAMES } from '../constants';
```
Comment (usavkov-epam) [inline src/TitleBindPieces/TitleBindPiecesCreateItemForm/TitleBindPiecesCreateItemForm.js:14]: plz look at the `PIECE_FORM_FIELD_NAMES`

File: src/TitleBindPieces/TitleBindPiecesCreateItemForm/TitleBindPiecesCreateItemForm.js [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+          label={<FormattedMessage id="ui-receiving.piece.permanentLocation" />}
+          name={TITLE_BIND_PIECES_FILED_NAMES.permanentLocation}
+          id={TITLE_BIND_PIECES_FILED_NAMES.permanentLocation}
+          component={LocationSelection}
+          onSelect={onLocationSelected}
+          fullWidth
+          marginBottom0
+          required
+        />
+        <LocationLookup onLocationSelected={onLocationSelected} />
```
Comment (usavkov-epam) [inline src/TitleBindPieces/TitleBindPiecesCreateItemForm/TitleBindPiecesCreateItemForm.js:105]: Please use existing component from acq-components lib

File: src/TitleBindPieces/TitleBindPiecesList.js [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+  COLUMN_FORMATTER,
+  COLUMN_MAPPING,
+  VISIBLE_COLUMNS,
+} from './constants';
+
+export const TitleBindPiecesList = ({ fields, props: { toggleCheckedAll } }) => {
+  const intl = useIntl();
+
+  const field = fields.name;
+  const cellFormatters = useMemo(
```
Comment (usavkov-epam) [inline src/TitleBindPieces/TitleBindPiecesList.js:24]: please use our style pattern for hooks (described earlier)

File: src/TitleBindPieces/TitleBindPiecesList.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    },
+    [field, intl],
+  );
+
+  const isAllChecked = useMemo(() => fields.value.every(({ checked }) => !!checked), [fields.value]);
+  const toggleAll = useCallback(
+    () => {
+      toggleCheckedAll(!isAllChecked);
+    },
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (usavkov-epam) [inline src/TitleBindPieces/TitleBindPiecesList.js:46]: please remove it

File: src/TitleBindPieces/TitleBindPiecesList.js [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        <Checkbox
+          checked={isAllChecked}
+          data-test-unreceive-title-checked
+          onChange={toggleAll}
+          aria-label={intl.formatMessage({ id: 'ui-receiving.piece.actions.selectAll' })}
+        />
+      ),
+      ...COLUMN_MAPPING,
+    }),
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (usavkov-epam) [inline src/TitleBindPieces/TitleBindPiecesList.js:62]: please remove it (or get meaningful reason why the rule should be ignored)

File: src/TitleBindPieces/hooks/useLoanTypes/useLoanTypes.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+  };
+
+  const { isLoading, data = {} } = useQuery(
+    [namespace],
+    () => ky.get(LOAN_TYPES_API, { searchParams }).json(),
+    {
+      enabled: true,
+      ...options,
+    },
+  );
```
Comment (usavkov-epam) [inline src/TitleBindPieces/hooks/useLoanTypes/useLoanTypes.js:30]: Colleagues from other teams have asked to pass an object in such hooks, specifying the data of the fields, to facilitate the cognitive perception of the passed arguments. I suggest following this approach.

File: translations/ui-receiving/en.json [hunk 9]
```diff
@@ -200,6 +202,7 @@
   "piece.comment": "Comment",
   "piece.enumeration": "Enumeration",
   "piece.chronology": "Chronology",
+  "piece.yearCaption": "Caption",
```
Comment (usavkov-epam) [inline translations/ui-receiving/en.json:205]: use display summary instead, please

File: src/TitleBindPieces/hooks/useUnboundPieces/useUnboundPieces.js [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+import { getPieceStatusFromItem } from '../../../common/utils';
+
+const DEFAULT_DATA = [];
+
+export const useUnboundPieces = (titleId, poLineId, options = {}) => {
+  const { enabled = true, ...otherOptions } = options;
+  const ky = useOkapiKy();
+  const [namespace] = useNamespace({ key: 'unbound-pieces' });
+  const filterQuery = `titleId=${titleId} and poLineId==${poLineId} and receivingStatus==${PIECE_STATUS.received} and isBound==false`;
```
Comment (usavkov-epam) [inline src/TitleBindPieces/hooks/useUnboundPieces/useUnboundPieces.js:26]: Can a title have multiple PO lines?

File: src/TitleBindPieces/hooks/useUnboundPieces/useUnboundPieces.js [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+  const searchParams = {
+    limit: LIMIT_MAX,
+    query: `${filterQuery} sortby locationId`,
+  };
+
+  const {
+    data = DEFAULT_DATA,
+    isLoading,
+  } = useQuery(
```
Comment (usavkov-epam) [inline src/TitleBindPieces/hooks/useUnboundPieces/useUnboundPieces.js:36]: Colleagues from other teams have asked to pass an object in such hooks, specifying the data of the fields, to facilitate the cognitive perception of the passed arguments. I suggest following this approach.

File: src/TitleBindPieces/hooks/useUnboundPieces/useUnboundPieces.js [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+      const pieceItems = await batchRequest(
+        ({ params }) => ky.get(ITEMS_API, { searchParams: params }).json(),
+        itemIds,
+      ).then(responses => responses.flatMap(({ items }) => items));
+
+      const itemsMap = keyBy(pieceItems, 'id');
+
+      return pieces.map((piece) => ({
+        ...piece,
+        yearCaption: piece.itemId ? itemsMap[piece.itemId].yearCaption : undefined,
```
Comment (usavkov-epam) [inline src/TitleBindPieces/hooks/useUnboundPieces/useUnboundPieces.js:58]: Display summary field should be used instead

**Review by usavkov-epam — COMMENTED** [review]
Apart from the comments, please take a look at the sonar code smells

File: src/TitleBindPieces/TitleBindPieces.js [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+  onCancel,
+  paneTitle,
+  pristine,
+  submitting,
+  values,
+  onSubmit,
+  isLoading,
+}) => {
+  const disabled = useMemo(() => {
+    if (!isRequiredFieldsFilled(values)) {
```
Comment (usavkov-epam) [inline src/TitleBindPieces/TitleBindPieces.js:37]: Why don't you use `valid` value from the form state instead? If required fields are not populated, then `valid === false`

File: src/TitleBindPieces/TitleBindPieces.js [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+  const shortcuts = [
+    {
+      name: 'cancel',
+      shortcut: 'esc',
+      handler: handleKeyCommand(onCancel),
+    },
+  ];
+
+  return (
+    <form onSubmit={() => onSubmit(values)}>
```
Comment (usavkov-epam) [inline src/TitleBindPieces/TitleBindPieces.js:64]: why not `handleSubmit`?

File: src/TitleBindPieces/TitleBindPiecesContainer/TitleBindPiecesContainer.js [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+TitleBindPiecesContainer.propTypes = {
+  history: ReactRouterPropTypes.history.isRequired,
+  location: ReactRouterPropTypes.location.isRequired,
+  match: ReactRouterPropTypes.match.isRequired,
+  stripes: stripesShape.isRequired,
+};
+
+export default stripesConnect(TitleBindPiecesContainer);
```
Comment (usavkov-epam) [inline src/TitleBindPieces/TitleBindPiecesContainer/TitleBindPiecesContainer.js:149]: you don't need `stripesConnect`. If you need the `stripes` object use `useStripes` hook

File: src/TitleBindPieces/TitleBindPieces.js [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+  }, [values]);
+
+  const paneFooter = (
+    <FormFooter
+      handleSubmit={handleSubmit}
+      isSubmitDisabled={disabled}
+      label={<FormattedMessage id="ui-receiving.title.details.button.bind" />}
+      onCancel={onCancel}
+      pristine={pristine}
+      submitting={submitting || isLoading}
```
Comment (usavkov-epam) [inline src/TitleBindPieces/TitleBindPieces.js:51]: please pass here only `submitting` value as the name of this prop requires it. If you need to disable submit button use `isSubmitDisabled`

File: src/TitleBindPieces/constants.js
Comment (usavkov-epam) [inline src/TitleBindPieces/constants.js:1]: a lot of stuff in this file have been already introduced, please, reuse it:
https://github.com/folio-org/ui-receiving/blob/0a2a14ddafd11a2bf8299f50d8beb119216108ee/src/TitleDetails/constants

File: src/TitleBindPieces/hooks/useLoanTypes/useLoanTypes.js [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+  const { enabled = true, ...restOptions } = options;
+  const ky = useOkapiKy();
+  const [namespace] = useNamespace({ key: 'loan-types' });
+
+  const searchParams = {
+    limit: LIMIT_MAX,
+    query: 'cql.allRecords=1 sortby name',
+  };
+
+  const { isLoading, data = {} } = useQuery(
```
Comment (usavkov-epam) [inline src/TitleBindPieces/hooks/useLoanTypes/useLoanTypes.js:24]: plz use an object of params as the argument:
```
useQuery({
  queryKey: ...,
  queryFn: ...,
})
```

File: src/TitleBindPieces/hooks/useMaterialTypes/useMaterialTypes.js [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+  };
+
+  const { isLoading, data = {} } = useQuery(
+    [namespace],
+    () => ky.get(MATERIAL_TYPE_API, { searchParams }).json(),
+    {
+      enabled,
+      ...restOptions,
+    },
+  );
```
Comment (usavkov-epam) [inline src/TitleBindPieces/hooks/useMaterialTypes/useMaterialTypes.js:31]: plz use an object of params as the argument:
```
useQuery({
  queryKey: ...,
  queryFn: ...,
})
```

**Comment by alisher-epam** [comment]
@folio-org/fe-tl-reviewers please review ^^^

**Comment by alisher-epam** [comment]
@usavkov-epam

**Review by usavkov-epam — APPROVED** [review]

**Pull request merged**